File: frontend/lib/tracker.ts
import type { ApplicationStatus } from "@/lib/api"

const BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"

export type { ApplicationStatus }

export interface ApplicationSummary {
  id: string
  resume_record_id: string | null
  jd_title: string
  jd_company: string
  status: ApplicationStatus
  applied_date: string | null
  follow_up_date: string | null
  archived_at: string | null
  created_at: string
  updated_at: string
}

export interface ApplicationFunnelResponse {
  status_counts: Record<ApplicationStatus, number>
  active_total: number
  archived_total: number
  total: number
  tracker_active_limit: number | null
}

export interface InterviewRound {
  id: string
  round_number: number
  name: string
  format: string
  scheduled_at: string | null
  duration_minutes: number | null
  interviewers: string[]
  notes: string | null
  outcome: string | null
  created_at: string
}

export interface OfferDetail {
  id: string
  base_salary_usd: number | null
  bonus_usd: number | null
  equity_description: string | null
  sign_on_usd: number | null
  benefits: string | null
  location: string | null
  remote: boolean
  start_date: string | null
  response_deadline: string | null
  decision: string | null
  decision_notes: string | null
  created_at: string
}

export interface ApplicationAttachment {
  id: string
  filename: string
  content_type: string
  size_bytes: number
  uploaded_at: string
  download_url?: string
}

export interface TimelineEvent {
  type: string
  at: string
  [key: string]: unknown
}

export interface ApplicationDetail extends ApplicationSummary {
  notes: string | null
  contact_name: string | null
  contact_email: string | null
  job_url: string | null
  rejection_reason: string | null
  rejection_notes: string | null
  status_history: { status: string; at: string }[]
  interview_rounds: InterviewRound[]
  offer_detail: OfferDetail | null
  attachments: ApplicationAttachment[]
  timeline: TimelineEvent[]
  attachment_usage: {
    count: number
    total_bytes: number
    max_count: number
    max_file_bytes: number
    max_total_bytes: number
  }
}

export interface Reminder {
  id: string
  scheduled_at: string
  message: string
  status: string
}

export const PIPELINE_COLUMNS: { key: ApplicationStatus; label: string }[] = [
  { key: "draft", label: "Draft" },
  { key: "applied", label: "Applied" },
  { key: "interviewing", label: "Interviewing" },
  { key: "offer", label: "Offer" },
  { key: "accepted", label: "Accepted" },
  { key: "rejected", label: "Rejected" },
  { key: "withdrawn", label: "Withdrawn" },
]

/**
 * Structured error thrown for non-2xx tracker responses.  Callers can
 * branch on ``code`` (`tracker_limit_reached`, `duplicate_application`,
 * ...) and read the raw ``detail`` payload without re-parsing the
 * message string.
 */
export class TrackerApiError extends Error {
  readonly status: number
  readonly code: string | undefined
  readonly detail: Record<string, unknown> | undefined

  constructor(
    message: string,
    opts: {
      status: number
      code?: string
      detail?: Record<string, unknown>
    },
  ) {
    super(message)
    this.name = "TrackerApiError"
    this.status = opts.status
    this.code = opts.code
    this.detail = opts.detail
  }
}

async function authRequest<T>(
  path: string,
  token: string,
  init?: RequestInit,
): Promise<T> {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${token}`,
    ...(init?.headers as Record<string, string> | undefined),
  }
  if (!(init?.body instanceof FormData)) {
    headers["Content-Type"] = "application/json"
  }
  const res = await fetch(`${BASE}${path}`, { ...init, headers })
  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    const detail = body?.detail
    const isObjectDetail = typeof detail === "object" && detail !== null
    const code = isObjectDetail ? (detail.code as string | undefined) : undefined
    const message =
      isObjectDetail
        ? (detail.message as string | undefined) ??
          (detail.detail as string | undefined)
        : typeof detail === "string"
          ? detail
          : undefined
    throw new TrackerApiError(message ?? code ?? `HTTP ${res.status}`, {
      status: res.status,
      code,
      detail: isObjectDetail ? (detail as Record<string, unknown>) : undefined,
    })
  }
  if (res.status === 204) return undefined as T
  return res.json() as Promise<T>
}

export async function listApplications(
  token: string,
  options: { archived?: "true" | "false" | "all" } = {},
): Promise<ApplicationSummary[]> {
  const search = options.archived
    ? `?archived=${encodeURIComponent(options.archived)}`
    : ""
  return authRequest(`/api/applications${search}`, token)
}

export async function getApplicationFunnel(
  token: string,
): Promise<ApplicationFunnelResponse> {
  return authRequest("/api/applications/funnel", token)
}

export async function archiveApplication(
  token: string,
  id: string,
): Promise<ApplicationSummary> {
  return authRequest(`/api/applications/${id}/archive`, token, {
    method: "POST",
  })
}

export async function unarchiveApplication(
  token: string,
  id: string,
): Promise<ApplicationSummary> {
  return authRequest(`/api/applications/${id}/unarchive`, token, {
    method: "POST",
  })
}

export async function getApplication(
  token: string,
  id: string,
): Promise<ApplicationDetail> {
  return authRequest(`/api/applications/${id}`, token)
}

export async function createApplication(
  token: string,
  body: {
    resume_record_id?: string
    jd_title?: string
    jd_company?: string
    job_url?: string
    status?: ApplicationStatus
    /** Set true to bypass duplicate-application 409 checks. */
    confirm_add_duplicate?: boolean
  },
): Promise<ApplicationSummary> {
  return authRequest("/api/applications", token, {
    method: "POST",
    body: JSON.stringify(body),
  })
}

export async function patchApplication(
  token: string,
  id: string,
  body: Record<string, unknown>,
): Promise<ApplicationSummary> {
  return authRequest(`/api/applications/${id}`, token, {
    method: "PATCH",
    body: JSON.stringify(body),
  })
}

export async function addInterviewRound(
  token: string,
  applicationId: string,
  body: {
    name: string
    format: string
    scheduled_at?: string
    duration_minutes?: number
    interviewers?: string[]
    notes?: string
    outcome?: string
  },
): Promise<InterviewRound> {
  return authRequest(`/api/applications/${applicationId}/rounds`, token, {
    method: "POST",
    body: JSON.stringify(body),
  })
}

export async function upsertOffer(
  token: string,
  applicationId: string,
  body: Record<string, unknown>,
  method: "POST" | "PATCH" = "PATCH",
): Promise<OfferDetail> {
  return authRequest(`/api/applications/${applicationId}/offer`, token, {
    method,
    body: JSON.stringify(body),
  })
}

export async function uploadAttachment(
  token: string,
  applicationId: string,
  file: File,
): Promise<ApplicationAttachment> {
  const form = new FormData()
  form.append("file", file)
  return authRequest(`/api/applications/${applicationId}/attachments`, token, {
    method: "POST",
    body: form,
  })
}

export async function deleteAttachment(
  token: string,
  applicationId: string,
  attachmentId: string,
): Promise<void> {
  await authRequest(
    `/api/applications/${applicationId}/attachments/${attachmentId}`,
    token,
    { method: "DELETE" },
  )
}

export async function listReminders(
  token: string,
  applicationId: string,
): Promise<Reminder[]> {
  return authRequest(`/api/applications/${applicationId}/reminders`, token)
}

export async function createReminder(
  token: string,
  applicationId: string,
  body: { scheduled_at: string; message: string },
): Promise<Reminder> {
  return authRequest(`/api/applications/${applicationId}/reminders`, token, {
    method: "POST",
    body: JSON.stringify(body),
  })
}

export async function deleteReminder(
  token: string,
  applicationId: string,
  reminderId: string,
): Promise<void> {
  await authRequest(
    `/api/applications/${applicationId}/reminders/${reminderId}`,
    token,
    { method: "DELETE" },
  )
}

export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`
}

File: backend/app/models/tracker.py
"""Application tracker models (SYSTEM_DESIGN_PHASE_2 §19.4)."""

from __future__ import annotations

import enum
import uuid
from datetime import date, datetime, timezone
from typing import Any

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    text,
)
from sqlalchemy.dialects.postgresql import ENUM as PGEnum
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class ApplicationStatus(str, enum.Enum):
    draft = "draft"
    applied = "applied"
    interviewing = "interviewing"
    offer = "offer"
    accepted = "accepted"
    rejected = "rejected"
    withdrawn = "withdrawn"


class InterviewFormat(str, enum.Enum):
    phone = "phone"
    video = "video"
    onsite = "onsite"
    take_home = "take_home"
    other = "other"


class InterviewOutcome(str, enum.Enum):
    pending = "pending"
    passed = "passed"
    failed = "failed"
    no_show = "no_show"


class OfferDecision(str, enum.Enum):
    pending = "pending"
    accepted = "accepted"
    declined = "declined"


class RejectionReason(str, enum.Enum):
    ghosted = "ghosted"
    explicit_rejection = "explicit_rejection"
    position_filled = "position_filled"
    withdrew = "withdrew"
    other = "other"


def _pg_enum(enum_cls: type[enum.Enum], name: str) -> PGEnum:
    return PGEnum(
        enum_cls,
        name=name,
        create_type=False,
        values_callable=lambda e: [m.value for m in e],
    )


_APPLICATION_STATUS_PG = _pg_enum(ApplicationStatus, "application_status")
_INTERVIEW_FORMAT_PG = _pg_enum(InterviewFormat, "interview_format")
_INTERVIEW_OUTCOME_PG = _pg_enum(InterviewOutcome, "interview_outcome")
_OFFER_DECISION_PG = _pg_enum(OfferDecision, "offer_decision")
_REJECTION_REASON_PG = _pg_enum(RejectionReason, "rejection_reason")

MAX_ATTACHMENTS_PER_APPLICATION = 5
MAX_ATTACHMENT_BYTES = 5 * 1024 * 1024
MAX_ATTACHMENT_TOTAL_BYTES = 25 * 1024 * 1024


class Application(Base):
    __tablename__ = "applications"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    resume_record_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("resume_records.id", ondelete="SET NULL"),
        nullable=True,
    )
    jd_title: Mapped[str] = mapped_column(String(500), nullable=False, default="")
    jd_company: Mapped[str] = mapped_column(String(500), nullable=False, default="")
    status: Mapped[ApplicationStatus] = mapped_column(
        _APPLICATION_STATUS_PG,
        nullable=False,
        default=ApplicationStatus.draft,
        server_default=ApplicationStatus.draft.value,
    )
    applied_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    follow_up_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    contact_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    contact_email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    job_url: Mapped[str | None] = mapped_column(String(2048), nullable=True)
    rejection_reason: Mapped[RejectionReason | None] = mapped_column(
        _REJECTION_REASON_PG, nullable=True
    )
    rejection_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    status_history: Mapped[list[dict[str, Any]]] = mapped_column(
        JSONB, nullable=False, default=list, server_default=text("'[]'::jsonb")
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=_utcnow,
        server_default=text("now()"),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=_utcnow,
        server_default=text("now()"),
    )

    interview_rounds: Mapped[list["InterviewRound"]] = relationship(
        back_populates="application",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="InterviewRound.round_number",
    )
    offer_detail: Mapped["OfferDetail | None"] = relationship(
        back_populates="application",
        cascade="all, delete-orphan",
        passive_deletes=True,
        uselist=False,
    )
    attachments: Mapped[list["ApplicationAttachment"]] = relationship(
        back_populates="application",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="ApplicationAttachment.uploaded_at",
    )

    __table_args__ = (
        Index("ix_applications_user_status", "user_id", "status"),
        Index(
            "ix_applications_resume_record_id",
            "resume_record_id",
            unique=True,
            postgresql_where=text("resume_record_id IS NOT NULL"),
        ),
    )


class InterviewRound(Base):
    __tablename__ = "interview_rounds"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    application_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("applications.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    round_number: Mapped[int] = mapped_column(Integer, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    format: Mapped[InterviewFormat] = mapped_column(
        _INTERVIEW_FORMAT_PG, nullable=False
    )
    scheduled_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    duration_minutes: Mapped[int | None] = mapped_column(Integer, nullable=True)
    interviewers: Mapped[list[str]] = mapped_column(
        JSONB, nullable=False, default=list, server_default=text("'[]'::jsonb")
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    outcome: Mapped[InterviewOutcome | None] = mapped_column(
        _INTERVIEW_OUTCOME_PG, nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=_utcnow,
        server_default=text("now()"),
    )

    application: Mapped["Application"] = relationship(back_populates="interview_rounds")


class OfferDetail(Base):
    __tablename__ = "offer_details"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    application_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("applications.id", ondelete="CASCADE"),
        nullable=False,
        unique=True,
    )
    base_salary_usd: Mapped[int | None] = mapped_column(Integer, nullable=True)
    bonus_usd: Mapped[int | None] = mapped_column(Integer, nullable=True)
    equity_description: Mapped[str | None] = mapped_column(Text, nullable=True)
    sign_on_usd: Mapped[int | None] = mapped_column(Integer, nullable=True)
    benefits: Mapped[str | None] = mapped_column(Text, nullable=True)
    location: Mapped[str | None] = mapped_column(String(255), nullable=True)
    remote: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    start_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    response_deadline: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    decision: Mapped[OfferDecision | None] = mapped_column(
        _OFFER_DECISION_PG, nullable=True
    )
    decision_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=_utcnow,
        server_default=text("now()"),
    )

    application: Mapped["Application"] = relationship(back_populates="offer_detail")


class ApplicationAttachment(Base):
    __tablename__ = "application_attachments"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    application_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("applications.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    filename: Mapped[str] = mapped_column(String(512), nullable=False)
    content_type: Mapped[str] = mapped_column(String(128), nullable=False)
    size_bytes: Mapped[int] = mapped_column(Integer, nullable=False)
    s3_key: Mapped[str] = mapped_column(String(1024), nullable=False)
    uploaded_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=_utcnow,
        server_default=text("now()"),
    )

    application: Mapped["Application"] = relationship(back_populates="attachments")


__all__ = [
    "MAX_ATTACHMENT_BYTES",
    "MAX_ATTACHMENT_TOTAL_BYTES",
    "MAX_ATTACHMENTS_PER_APPLICATION",
    "Application",
    "ApplicationAttachment",
    "ApplicationStatus",
    "InterviewFormat",
    "InterviewOutcome",
    "InterviewRound",
    "OfferDecision",
    "OfferDetail",
    "RejectionReason",
]
